// iterable-operator/lib/es2015/middleware/drop-async.js
"use strict";
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.dropAsync = exports.InvalidArgumentError = void 0;
const utils_1 = require("../utils");
const error_1 = require("../error");
Object.defineProperty(exports, "InvalidArgumentError", { enumerable: true, get: function () { return error_1.InvalidArgumentError; } });
function dropAsync(iterable, count) {
    if (count < 0)
        throw new error_1.InvalidArgumentError('count', '>= 0');
    if (count === 0)
        return utils_1.copyAsyncIterable(iterable);
    return (function () {
        return __asyncGenerator(this, arguments, function* () {
            const iterator = iterable[Symbol.asyncIterator]();
            let result;
            while (result = yield __await(iterator.next()), !result.done) {
                if (count <= 0)
                    break;
                count--;
            }
            while (!result.done) {
                yield yield __await(result.value);
                result = yield __await(iterator.next());
            }
        });
    })();
}
exports.dropAsync = dropAsync;
