// iterable-operator/lib/es2015/middleware/flatten-by-async.js
"use strict";
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncDelegator = (this && this.__asyncDelegator) || function (o) {
    var i, p;
    return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
    function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
};
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.flattenByAsync = void 0;
const types_1 = require("@blackglory/types");
function flattenByAsync(iterable, fn) {
    if (types_1.isAsyncIterable(iterable)) {
        return flattenByAsyncIterable(iterable);
    }
    else {
        return flattenByIterable(iterable);
    }
    function flattenByAsyncIterable(iterable) {
        return __asyncGenerator(this, arguments, function* flattenByAsyncIterable_1() {
            var e_1, _a;
            const level = 1;
            try {
                for (var iterable_1 = __asyncValues(iterable), iterable_1_1; iterable_1_1 = yield __await(iterable_1.next()), !iterable_1_1.done;) {
                    const element = iterable_1_1.value;
                    if (isFiniteIterable(element) && (yield __await(fn(element, level)))) {
                        yield __await(yield* __asyncDelegator(__asyncValues(flatten(element, level + 1))));
                    }
                    else {
                        yield yield __await(element);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (iterable_1_1 && !iterable_1_1.done && (_a = iterable_1.return)) yield __await(_a.call(iterable_1));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    function flattenByIterable(iterable) {
        return flatten(iterable, 1);
    }
    function flatten(iterable, level) {
        return __asyncGenerator(this, arguments, function* flatten_1() {
            for (const element of iterable) {
                if (isFiniteIterable(element) && (yield __await(fn(element, level)))) {
                    yield __await(yield* __asyncDelegator(__asyncValues(flatten(element, level + 1))));
                }
                else {
                    yield yield __await(element);
                }
            }
        });
    }
}
exports.flattenByAsync = flattenByAsync;
function isFiniteIterable(val) {
    return types_1.isIterable(val) && types_1.isntChar(val);
}
