// extra-dom/lib/es2018/node-constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeConstants = void 0;
const nodeExists = 'Node' in globalThis;
var NodeConstants;
(function (NodeConstants) {
    NodeConstants[NodeConstants["ELEMENT_NODE"] = nodeExists ? Node.ELEMENT_NODE : 1] = "ELEMENT_NODE";
    NodeConstants[NodeConstants["ATTRIBUTE_NODE"] = nodeExists ? Node.ATTRIBUTE_NODE : 2] = "ATTRIBUTE_NODE";
    NodeConstants[NodeConstants["TEXT_NODE"] = nodeExists ? Node.TEXT_NODE : 3] = "TEXT_NODE";
    NodeConstants[NodeConstants["CDATA_SECTION_NODE"] = nodeExists ? Node.CDATA_SECTION_NODE : 4] = "CDATA_SECTION_NODE";
    NodeConstants[NodeConstants["ENTITY_REFERENCE_NODE"] = nodeExists ? Node.ENTITY_REFERENCE_NODE : 5] = "ENTITY_REFERENCE_NODE";
    NodeConstants[NodeConstants["ENTITY_NODE"] = nodeExists ? Node.ENTITY_NODE : 6] = "ENTITY_NODE";
    NodeConstants[NodeConstants["PROCESSING_INSTRUCTION_NODE"] = nodeExists ? Node.PROCESSING_INSTRUCTION_NODE : 7] = "PROCESSING_INSTRUCTION_NODE";
    NodeConstants[NodeConstants["COMMENT_NODE"] = nodeExists ? Node.COMMENT_NODE : 8] = "COMMENT_NODE";
    NodeConstants[NodeConstants["DOCUMENT_NODE"] = nodeExists ? Node.DOCUMENT_NODE : 9] = "DOCUMENT_NODE";
    NodeConstants[NodeConstants["DOCUMENT_TYPE_NODE"] = nodeExists ? Node.DOCUMENT_TYPE_NODE : 10] = "DOCUMENT_TYPE_NODE";
    NodeConstants[NodeConstants["DOCUMENT_FRAGMENT_NODE"] = nodeExists ? Node.DOCUMENT_FRAGMENT_NODE : 11] = "DOCUMENT_FRAGMENT_NODE";
    NodeConstants[NodeConstants["NOTATION_NODE"] = nodeExists ? Node.NOTATION_NODE : 12] = "NOTATION_NODE";
})(NodeConstants = exports.NodeConstants || (exports.NodeConstants = {}));
